addons/web_tour/static/src/js/tour_service.js

```diff
@@ -140,18 +146,37 @@ return session.is_bound.then(function () {
         });
 
         // Now that the observer is configured, we have to start it when needed.
+        const observerOptions = {
+            attributes: true,
+            childList: true,
+            subtree: true,
+            attributeOldValue: true,
+            characterData: true,
+        };
+
         var start_service = (function () {
             return function (observe) {
                 return new Promise(function (resolve, reject) {
                     tour_manager._register_all(observe).then(function () {
                         if (observe) {
-                            observer.observe(document.body, {
-                                attributes: true,
-                                childList: true,
-                                subtree: true,
-                                attributeOldValue: true,
-                                characterData: true,
+                            observer.observe(document.body, observerOptions);
+
+                            // If an iframe is added during the tour, its DOM
+                            // mutations should also be observed to update the
+                            // tour manager.
+                            const iframeObserver = new MutationObserver(mutations => {
+                                const iframeMutation = mutations.filter(mutation => Array.from(mutation.addedNodes).some(node => {
+                                    const isElement = node.nodeType === 1;
+                                    return isElement && node.classList.contains(iframeContainer);
+                                }))[0];
+                                if (iframeMutation) {
+                                    const iframeEl = iframeMutation.target.querySelector('iframe');
+                                    iframeEl.addEventListener('load', () => {
+                                        observer.observe(iframeEl.contentDocument.body, observerOptions);
+                                    });
+                                }
                             });
+                            iframeObserver.observe(document.body, { childList: true, subtree: true });
```

Comment:
Can't we just consider all iframes? If not, nothing is blocking you from declaring the iframe class names in their related apps like you suggests to do for master only, you can do it directly here, no?

Maybe we could merge that in 15.0 too? It's not a critical bug to solve before that since we probably don't have tours checking iframes but we could give the community the opportunity starting from 15.0.

Reply:
Considering all iframes seemed to be the ideal fix (the fix is stable and adding a new client action using an iframe is easier), but there were two issues that made me go for the whitelist: 
- the mutation will not report the addition of the iframe directly, but its container (`o_website_editor` for the client action, `wysiwyg_iframe` for mass_mailing), but we can also, for each added node, query to find an iframe. 
I was apprehensive of the performances but maybe it's not a problem ?
- we only want to monitor the iframes that the user is interacting with (for now, only the website and the snippet editor), and in the end, it seemed permissive to listen on all the iframes (for example, if the user has a youtube video on his webpage and runs a tour)
What do you think ? 
  
In master, to define the iframe container classes I defined a new class TourService in a new `web_tour.TourService` that will be patched from mass_mailing and website, and I thought that it was not possible to do that in stable, but maybe there is another way to do that ?

Reply:
@younn-o 

> [...] but we can also, for each added node, query to find an iframe

Is that not what you already do?

> we only want to monitor the iframes that the user is interacting with [...] or example, if the user has a youtube video on his webpage and runs a tour

I thought about that and it did not seem like a big deal but indeed, the youtube video is a valid point, let's go with the whitelist indeed.

> In master, to define the iframe container classes I defined a new class TourService in a new web_tour.TourService that will be patched from mass_mailing and website

Well, you can do the same in stable: not converting the whole tour_service.js into a class of course, but just adding a new Class that is instantiated around here or even a simple function, that you then patch in other apps. But maybe there is even another way: what about just considering the iframes that have the o_iframe class (if it does not already exist for something, maybe "o_wt_iframe" at worst (wt for web_tour))? Then if as I suspect all iframes are built via JS, it is stable to add that class on them. And this would allow the community to not have to patch Classes and stuff, just adding a class on the iframe.